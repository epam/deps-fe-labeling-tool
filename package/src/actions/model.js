
import { batch } from 'react-redux'
import { createAction } from 'redux-actions'
import { LabelType, LABEL_TYPE_NAME } from '@/models/Label'
import { Markup } from '@/models/Markup'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector, fieldsToDeleteSelector } from '@/selectors/model'
import {
  updateLabelsWithSettings,
  updateTablesWithSettings
} from './markup'

const FEATURE_NAME = 'MODEL'

const storeFields = createAction(
  `${FEATURE_NAME}/STORE`
)

const storeInitialFields = createAction(
  `${FEATURE_NAME}/STORE_INITIAL_FIELDS`
)

const updateInitialFields = createAction(
  `${FEATURE_NAME}/UPDATE_INITIAL_FIELDS`
)

const enableDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/ENABLE_DELETE_FIELD_MODE`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const updateFieldsToDelete = createAction(
  `${FEATURE_NAME}/UPDATE_FIELDS_TO_DELETE`
)

const cancelDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/CANCEL_DELETE_FIELDS_MODE`
)

const confirmFieldsDeletion = () => (dispatch, getState) => {
  const state = getState()
  const markup = markupSelector(state)
  const fields = fieldsSelector(state)
  const fieldsToDelete = fieldsToDeleteSelector(state)
  const newFields = fields.filter((field) => fieldsToDelete.every((f) => f.code !== field.code))
  const assignedObjects = markup
    ? Markup.getAllObjects(markup).filter((markupObject) => fieldsToDelete.find((field) => field.code === markupObject.fieldCode))
    : []

  batch(() => {
    assignedObjects.forEach((markupObject) => {
      if (markupObject.typeName === LABEL_TYPE_NAME) {
        dispatch(updateLabelsWithSettings({
          ...markupObject,
          fieldCode: '',
          type: LabelType.UNASSIGNED,
          index: undefined
        }))
      } else {
        dispatch(updateTablesWithSettings({
          ...markupObject,
          fieldCode: '',
          index: undefined
        }))
      }
    })
    dispatch(storeFields(newFields))
    dispatch(cancelDeleteFieldsMode())
  })
}

export {
  enableDeleteFieldsMode,
  resetDefault,
  storeFields,
  storeInitialFields,
  updateFieldsToDelete,
  updateInitialFields,
  confirmFieldsDeletion,
  cancelDeleteFieldsMode
}
