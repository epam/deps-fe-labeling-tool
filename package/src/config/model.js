import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { PairFieldMeta, ListFieldMeta, EnumFieldMeta } from '@/models/FieldMeta'

const fields = [
  new Field(
    'dates',
    'Dates',
    FieldType.DATE
  ),
  new Field(
    'firstName',
    'First Name',
    FieldType.STRING
  ),
  new Field(
    'lastName',
    'Last Name',
    FieldType.STRING,
    undefined,
    true
  ),
  new Field(
    'companyName',
    'Company Name',
    FieldType.STRING,
    undefined,
    true
  ),
  new Field(
    'city',
    'City',
    FieldType.ENUM,
    new EnumFieldMeta([
      'Anytown',
      'Sometown',
      'San Diego',
      'Los Angeles'
    ]),
    true
  ),
  new Field(
    'countryStates',
    'Country States',
    FieldType.LIST,
    new ListFieldMeta(FieldType.STRING),
    true
  ),
  new Field(
    'dateMatching',
    'Date Matching',
    FieldType.LIST,
    new ListFieldMeta(FieldType.CHECKMARK),
    true
  ),
  new Field(
    'date',
    'Date',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING)
  ),
  new Field(
    'zipCodes',
    'Zip Codes',
    FieldType.LIST,
    new ListFieldMeta(
      FieldType.PAIR,
      new PairFieldMeta(FieldType.STRING, FieldType.STRING)
    ),
    true
  ),
  new Field(
    'homeAddress',
    'Home Address',
    FieldType.STRING
  ),
  new Field(
    'businessAddress',
    'Business Address',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING),
    true
  ),
  new Field(
    'taxpayerSpousesInfo',
    'Taxpayer Spouse\'s Info',
    FieldType.TABLE
  ),
  new Field(
    'taxpayerInfo',
    'Taxpayer Info',
    FieldType.TABLE,
    undefined,
    true
  ),
  new Field(
    'tables',
    'Tables',
    FieldType.LIST,
    new ListFieldMeta(FieldType.TABLE)
  ),
  new Field(
    'uniqueIdentifier',
    'Unique Identifier',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING)
  ),
  new Field(
    'signatory',
    'Signatory',
    FieldType.CHECKMARK
  ),
  new Field(
    'returnTranscript',
    'Return Transcript',
    FieldType.CHECKMARK
  ),
  new Field(
    'formNumbers',
    'Form Numbers',
    FieldType.LIST,
    new ListFieldMeta(
      FieldType.ENUM,
      new EnumFieldMeta([
        '1098',
        '5498',
        '1040'
      ])
    )
  )
]

export {
  fields
}
