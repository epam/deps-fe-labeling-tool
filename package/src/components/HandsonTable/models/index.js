
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
