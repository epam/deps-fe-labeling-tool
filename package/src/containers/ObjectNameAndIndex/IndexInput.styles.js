import styled from 'styled-components'
import { Input } from '@/components/Input'
import { COLORS } from '@/theme/theme.default'

const StyledNumericInput = styled(Input.Numeric)`
  flex: 0 1 6.4rem;
  background-color: ${COLORS.PRIMARY_3};

  & input {
    padding: 0 0.8rem;
  }

  & input:disabled{
    background-color: ${COLORS.PRIMARY_3};
  }
`

export {
  StyledNumericInput as NumericInput
}
