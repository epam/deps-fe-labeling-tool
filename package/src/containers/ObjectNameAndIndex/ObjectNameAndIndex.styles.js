import styled from 'styled-components'
import { Typography } from '@/components/Typography'
import { COLORS } from '@/theme/theme.default'

const { Paragraph } = Typography

const ObjectName = styled(Paragraph)`
  max-width: 190px;
  height: 2.4rem;
  font-size: 1.2rem;
  line-height: 1.6rem;
  color: ${COLORS.PRIMARY_4};
  padding: 0.4rem 0 0 0.6rem;
  margin-right: 1rem;
  border-left: 2px solid ${(props) => props.color};
`

const ObjectType = styled.span`
  font-size: 1rem;
  line-height: 1.2rem;
  color: ${COLORS.GRAYSCALE_5};
`

const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  height: 2.4rem;
`

export {
  ObjectName,
  ObjectType,
  Wrapper
}
