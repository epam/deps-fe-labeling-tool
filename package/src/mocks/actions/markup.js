const addLabels = jest.fn((labels) => ({
  type: 'MOCK_ADD_LABELS_ACTION',
  payload: labels
}))

const addTables = jest.fn((tables) => ({
  type: 'MOCK_ADD_TABLES_ACTION',
  payload: tables
}))

const removeLabels = jest.fn((labels) => ({
  type: 'MOCK_REMOVE_LABELS_ACTION',
  payload: labels
}))

const updateLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_LABELS_ACTION',
  payload: labels
}))

const updateAllLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_ALL_LABELS_ACTION',
  payload: labels
}))

const updateTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_TABLES_ACTION',
  payload: tables
}))

const updateAllTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_ALL_TABLES_ACTION',
  payload: tables
}))

const selectLabels = jest.fn((labels) => ({
  type: 'MOCK_SELECT_LABELS_ACTION',
  payload: labels
}))

const selectTables = jest.fn((tables) => ({
  type: 'MOCK_SELECT_TABLES_ACTION',
  payload: tables
}))

const clearSelection = jest.fn(() => ({
  type: 'MOCK_CLEAR_SELECTION_ACTION'
}))

const removeTables = jest.fn((tables) => ({
  type: 'MOCK_REMOVE_TABLES_ACTION',
  payload: tables
}))

const importMarkup = jest.fn(() => ({
  type: 'MOCK_IMPORT_MARKUP'
}))

const exportMarkup = jest.fn(() => ({
  type: 'MOCK_EXPORT_MARKUP'
}))

const undo = jest.fn(() => ({
  type: 'MOCK_UNDO_ACTION'
}))

const redo = jest.fn(() => ({
  type: 'MOCK_REDO_ACTION'
}))

const updateInitialMarkup = jest.fn(() => ({
  type: 'MOCK_UPDATE_INITIAL_MARKUP_ACTION'
}))

const storeMarkup = jest.fn((markup) => ({
  type: 'MOCK_STORE_MARKUP_ACTION',
  payload: markup
}))

const storeAssignedMarkup = jest.fn((markup) => ({
  type: 'MOCK_UPDATE_MARKUP_ACTION',
  payload: markup
}))

const updatePageMarkup = jest.fn(() => ({
  type: 'UPDATE_PAGE_MARKUP'
}))

const reset = jest.fn(() => ({
  type: 'RESET_MARKUP_ACTION'
}))

const copyMarkup = jest.fn(() => ({
  type: 'COPY_MARKUP_ACTION'
}))

const pasteMarkup = jest.fn(() => ({
  type: 'PASTE_MARKUP_ACTION'
}))

const updateLabelsWithSettings = jest.fn(() => ({
  type: 'UPDATE_LABELS_WITH_SETTINGS'
}))

const updateTablesWithSettings = jest.fn(() => ({
  type: 'UPDATE_TABLES_WITH_SETTINGS'
}))

const mockMarkupActions = {
  addLabels,
  addTables,
  copyMarkup,
  pasteMarkup,
  removeLabels,
  reset,
  updateLabels,
  updateAllLabels,
  updateTables,
  updateAllTables,
  selectLabels,
  selectTables,
  clearSelection,
  removeTables,
  importMarkup,
  exportMarkup,
  undo,
  updateInitialMarkup,
  storeMarkup,
  storeAssignedMarkup,
  updatePageMarkup,
  redo,
  updateLabelsWithSettings,
  updateTablesWithSettings
}

export {
  mockMarkupActions
}
