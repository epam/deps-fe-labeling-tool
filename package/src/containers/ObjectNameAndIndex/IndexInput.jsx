import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { deleteTemporaryFieldIndex } from '@/actions/ui'
import { labelShape } from '@/models/Label'
import { tableShape } from '@/models/Table'
import { NumericInput } from './IndexInput.styles'

const INDEX_MIN_VALUE = 1

const IndexInput = ({
  onChange,
  object,
  className,
  deleteTemporaryFieldIndex
}) => {
  const onIndexChange = useCallback(
    (value) => {
      if (object.index === value) {
        return
      }

      deleteTemporaryFieldIndex({
        code: object.fieldCode,
        index: value
      })

      onChange(
        {
          ...object,
          index: value
        }
      )
    },
    [
      object,
      onChange,
      deleteTemporaryFieldIndex
    ]
  )

  return (
    <NumericInput
      className={className}
      min={INDEX_MIN_VALUE}
      value={object.index}
      onChange={onIndexChange}
    />
  )
}

IndexInput.propTypes = {
  className: PropTypes.string,
  deleteTemporaryFieldIndex: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  object: PropTypes.oneOfType([
    labelShape,
    tableShape
  ]).isRequired
}

const mapDispatchToProps = {
  deleteTemporaryFieldIndex
}

const ConnectedComponent = connect(null, mapDispatchToProps)(IndexInput)

export {
  ConnectedComponent as IndexInput
}
