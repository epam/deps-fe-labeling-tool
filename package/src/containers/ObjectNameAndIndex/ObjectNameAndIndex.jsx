import React, { useMemo } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Tooltip } from '@/components/Tooltip'
import { FieldType } from '@/enums/FieldType'
import { fieldShape } from '@/models/Field'
import { labelShape, LabelType } from '@/models/Label'
import { Markup, markupShape } from '@/models/Markup'
import { tableShape, TABLE_TYPE_NAME } from '@/models/Table'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector } from '@/selectors/model'
import { COLORS } from '@/theme/theme.default'
import { IndexInput } from './IndexInput'
import {
  ObjectName,
  Wrapper,
  ObjectType
} from './ObjectNameAndIndex.styles'

const PAIR_NAME_SEPARATOR = '.'
const RESOURCE_UNASSIGNED = 'Unassigned'

const OBJECT_TYPE_TO_COLOR = {
  [LabelType.CHECKMARK]: COLORS.LABEL_CHECKMARK_BORDER,
  [LabelType.KEY]: COLORS.LABEL_KEY_BORDER,
  [LabelType.VALUE]: COLORS.LABEL_VALUE_BORDER,
  [LabelType.STRING]: COLORS.LABEL_STRING_BORDER,
  [LabelType.ENUM]: COLORS.LABEL_ENUM_BORDER,
  [LabelType.DATE]: COLORS.LABEL_DATE_BORDER,
  [TABLE_TYPE_NAME]: COLORS.TABLE_BORDER
}

const ObjectNameAndIndex = ({ object, onChange, fields, markup }) => {
  const indexAvailable = useMemo(
    () => {
      const { fieldType } = fields.find((f) => f.code === object.fieldCode) ?? {}
      return fieldType === FieldType.LIST
    },
    [object, fields]
  )

  const getKeyValuePairName = (name, type) => (
    <span>
      {`${name}${PAIR_NAME_SEPARATOR}`}
      <ObjectType>
        {type}
      </ObjectType>
    </span>
  )

  const getObjectColor = () => {
    if (!object.fieldCode) {
      return COLORS.MARKUP_OBJECT_UNASSIGNED_BORDER
    }

    return OBJECT_TYPE_TO_COLOR[object.type ?? object.typeName]
  }

  const getUnassignedObjectName = () => {
    const allMarkupObjects = Markup.getAllObjects(markup)

    const filteredMarkupObjects = allMarkupObjects.filter((markupObject) => (
      !markupObject.fieldCode &&
      markupObject.typeName === object.typeName
    ))

    const objectIndex = filteredMarkupObjects.findIndex((markupObject) => (
      markupObject.uid === object.uid
    ))

    return `${RESOURCE_UNASSIGNED} ${object.typeName} ${objectIndex + 1}`
  }

  const getObjectName = () => {
    const objectName = fields.find((field) => field.code === object.fieldCode)?.name

    if (object.fieldCode === '') {
      return getUnassignedObjectName()
    }

    if (object.type === LabelType.KEY || object.type === LabelType.VALUE) {
      return getKeyValuePairName(objectName, object.type)
    }

    return objectName
  }

  return (
    <Wrapper>
      <ObjectName
        ellipsis={
          {
            tooltip: (
              <Tooltip
                title={object.typeName}
              >
                {getObjectName()}
              </Tooltip>
            )
          }
        }
        color={getObjectColor()}
      >
        {getObjectName()}
      </ObjectName>
      {
        indexAvailable &&
        (
          <IndexInput
            onChange={onChange}
            object={object}
            fields={fields}
          />
        )
      }
    </Wrapper>
  )
}

ObjectNameAndIndex.propTypes = {
  object: PropTypes.oneOfType([
    labelShape, tableShape
  ]).isRequired,
  onChange: PropTypes.func.isRequired,
  fields: PropTypes.arrayOf(fieldShape).isRequired,
  markup: markupShape.isRequired
}

const mapStateToProps = (state) => ({
  fields: fieldsSelector(state),
  markup: markupSelector(state)
})

const ConnectedComponent = connect(mapStateToProps)(ObjectNameAndIndex)

export {
  ConnectedComponent as ObjectNameAndIndex
}
