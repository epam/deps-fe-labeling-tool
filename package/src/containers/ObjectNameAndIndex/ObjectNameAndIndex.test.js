import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockMarkupSelectors } from '@/mocks/selectors/markup'
import { mockModelSelectors } from '@/mocks/selectors/model'
import React from 'react'
import { shallow } from 'enzyme'
import { Label, LabelType } from '@/models/Label'
import { ObjectNameAndIndex } from './ObjectNameAndIndex'

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/selectors/markup', () => mockMarkupSelectors)
jest.mock('@/selectors/model', () => mockModelSelectors)

jest.mock('@/components/Tooltip', () => mockComponent('Tooltip'))
jest.mock('./IndexInput', () => mockComponent('IndexInput'))

const {
  WrappedComponent,
  mapStateToProps
} = ObjectNameAndIndex

describe('Container: ObjectNameAndIndex', () => {
  describe('mapStateToProps', () => {
    it('should pass fields from modelSelector and pass as result to WrappedComponent', () => {
      const { props } = mapStateToProps()
      expect(props.fields).toEqual(mockModelSelectors.fieldsSelector())
    })

    it('should pass markup from markupSelector and pass as result to WrappedComponent', () => {
      const { props } = mapStateToProps()
      expect(props.markup).toEqual(mockMarkupSelectors.markupSelector())
    })
  })

  describe('ConnectedComponent', () => {
    let wrapper, defaultProps

    beforeEach(() => {
      defaultProps = {
        ...mapStateToProps().props,
        object: new Label(1, 2, 3, 4, 'testCode', undefined, LabelType.STRING, ''),
        onChange: jest.fn()
      }

      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render layout correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should correctly render label with type key', () => {
      defaultProps.object = new Label(1, 2, 3, 4, 'testKey', undefined, LabelType.KEY, '')
      wrapper.setProps(defaultProps)
      expect(wrapper).toMatchSnapshot()
    })

    it('should correctly render label with type checkmark', () => {
      defaultProps.object = new Label(1, 2, 3, 4, 'testKey', undefined, LabelType.CHECKMARK, false)
      wrapper.setProps(defaultProps)
      expect(wrapper).toMatchSnapshot()
    })

    it('should correctly render unassignedLabel', () => {
      defaultProps.object = new Label(1, 2, 3, 4, '', undefined, LabelType.UNASSIGNED, '')
      wrapper.setProps(defaultProps)
      expect(wrapper).toMatchSnapshot()
    })
  })
})
