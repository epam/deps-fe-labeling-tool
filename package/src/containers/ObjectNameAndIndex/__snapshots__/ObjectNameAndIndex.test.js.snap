// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: ObjectNameAndIndex ConnectedComponent should correctly render label with type checkmark 1`] = `
<ObjectNameAndIndexstyles__Wrapper>
  <ObjectNameAndIndexstyles__ObjectName
    color="rgba(141, 180, 31, 1)"
    ellipsis={
      Object {
        "tooltip": <Tooltip
          title="label"
        />,
      }
    }
  />
</ObjectNameAndIndexstyles__Wrapper>
`;

exports[`Container: ObjectNameAndIndex ConnectedComponent should correctly render label with type key 1`] = `
<ObjectNameAndIndexstyles__Wrapper>
  <ObjectNameAndIndexstyles__ObjectName
    color="#3980E8"
    ellipsis={
      Object {
        "tooltip": <Tooltip
          title="label"
        >
          <span>
            undefined.
            <ObjectNameAndIndexstyles__ObjectType>
              key
            </ObjectNameAndIndexstyles__ObjectType>
          </span>
        </Tooltip>,
      }
    }
  >
    <span>
      undefined.
      <ObjectNameAndIndexstyles__ObjectType>
        key
      </ObjectNameAndIndexstyles__ObjectType>
    </span>
  </ObjectNameAndIndexstyles__ObjectName>
</ObjectNameAndIndexstyles__Wrapper>
`;

exports[`Container: ObjectNameAndIndex ConnectedComponent should correctly render unassignedLabel 1`] = `
<ObjectNameAndIndexstyles__Wrapper>
  <ObjectNameAndIndexstyles__ObjectName
    color="#FF5645"
    ellipsis={
      Object {
        "tooltip": <Tooltip
          title="label"
        >
          Unassigned label 0
        </Tooltip>,
      }
    }
  >
    Unassigned label 0
  </ObjectNameAndIndexstyles__ObjectName>
</ObjectNameAndIndexstyles__Wrapper>
`;

exports[`Container: ObjectNameAndIndex ConnectedComponent should render layout correctly 1`] = `
<ObjectNameAndIndexstyles__Wrapper>
  <ObjectNameAndIndexstyles__ObjectName
    color="#FFB357"
    ellipsis={
      Object {
        "tooltip": <Tooltip
          title="label"
        />,
      }
    }
  />
</ObjectNameAndIndexstyles__Wrapper>
`;
